File: src/tui/data-source.ts
import {
  addTenant as configAddTenant,
  removeTenant as configRemoveTenant,
  loadConfig,
} from "../config/index.ts";
import type { BuildReport } from "../engine/builder.ts";
import type { ArtifactView, ListFilter } from "../engine/reader.ts";
import { type TenantInfo, TenantManager } from "../engine/tenant.ts";
import { err, ok, type Result } from "../lib/result.ts";
import type { SearchResult } from "../storage/interface.ts";

export type DataSourceError = { type: "data-source"; message: string };

export type DataSource = {
  get(tenant: string, slug: string): Promise<Result<ArtifactView | null, DataSourceError>>;
  search(tenant: string, query: string): Promise<Result<SearchResult[], DataSourceError>>;
  traverse(
    tenant: string,
    from: string,
    relType?: string,
    depth?: number,
  ): Promise<Result<ArtifactView[], DataSourceError>>;
  list(tenant: string, filter?: ListFilter): Promise<Result<ArtifactView[], DataSourceError>>;
  listTenants(): Promise<Result<TenantInfo[], DataSourceError>>;
  reload(tenant: string): Promise<Result<BuildReport, DataSourceError>>;
  addKb(name: string, sourceDir: string): Promise<Result<void, DataSourceError>>;
  removeKb(name: string): Promise<Result<void, DataSourceError>>;
  mode: "daemon" | "standalone";
};

function dsErr(message: string): DataSourceError {
  return { type: "data-source", message };
}

export function createReaderDataSource(tm: TenantManager): DataSource {
  return {
    mode: "standalone",

    async get(tenant, slug) {
      const r = tm.getReader(tenant);
      if (!r.ok) return err(dsErr(r.error.message));
      const result = r.value.get(slug);
      if (!result.ok) return err(dsErr(result.error.message));
      return ok(result.value);
    },

    async search(tenant, query) {
      const r = tm.getReader(tenant);
      if (!r.ok) return err(dsErr(r.error.message));
      const result = await r.value.search(query);
      if (!result.ok) return err(dsErr(result.error.message));
      return ok(result.value);
    },

    async traverse(tenant, from, relType, depth) {
      const r = tm.getReader(tenant);
      if (!r.ok) return err(dsErr(r.error.message));
      const result = r.value.traverse(from, relType, depth);
      if (!result.ok) return err(dsErr(result.error.message));
      return ok(result.value);
    },

    async list(tenant, filter) {
      const r = tm.getReader(tenant);
      if (!r.ok) return err(dsErr(r.error.message));
      const result = r.value.list(filter);
      if (!result.ok) return err(dsErr(result.error.message));
      return ok(result.value);
    },

    async listTenants() {
      return ok(tm.listTenants());
    },

    async reload(tenant) {
      const result = await tm.reload(tenant);
      if (!result.ok) return err(dsErr(result.error.message));
      return ok(result.value);
    },

    async addKb(name, sourceDir) {
      const mountResult = await tm.mount({ name, sourceDir });
      if (!mountResult.ok) return err(dsErr(mountResult.error.message));
      const configResult = await configAddTenant(name, sourceDir);
      if (!configResult.ok) {
        tm.unmount(name);
        return err(dsErr(configResult.error.message));
      }
      return ok(undefined);
    },

    async removeKb(name) {
      const unmountResult = tm.unmount(name);
      if (!unmountResult.ok) return err(dsErr(unmountResult.error.message));
      const configResult = await configRemoveTenant(name);
      if (!configResult.ok) return err(dsErr(configResult.error.message));
      return ok(undefined);
    },
  };
}

export async function createStandaloneFromConfig(): Promise<Result<DataSource, DataSourceError>> {
  const tm = new TenantManager();
  await tm.initEmbedder();

  const configResult = await loadConfig();
  if (configResult.ok) {
    for (const [name, dir] of Object.entries(configResult.value.tenants)) {
      await tm.mount({ name, sourceDir: dir });
    }
  }

  return ok(createReaderDataSource(tm));
}

export function createHttpDataSource(port: string): DataSource {
  const baseUrl = `http://localhost:${port}`;

  async function fetchJson<T>(url: string, method = "GET"): Promise<Result<T, DataSourceError>> {
    try {
      const res = await fetch(url, { method });
      if (!res.ok) {
        const body = (await res.json()) as { error?: string };
        return err(dsErr(body.error ?? `HTTP ${res.status}`));
      }
      return ok((await res.json()) as T);
    } catch (e) {
      return err(dsErr(e instanceof Error ? e.message : "Network error"));
    }
  }

  return {
    mode: "daemon",

    async get(tenant, slug) {
      const parts = slug.split("#");
      const url =
        parts.length > 1
          ? `${baseUrl}/v1/${tenant}/get/${parts[0]}/${parts[1]}`
          : `${baseUrl}/v1/${tenant}/get/${slug}`;
      try {
        const res = await fetch(url);
        if (res.status === 404) return ok(null);
        if (!res.ok) {
          const body = (await res.json()) as { error?: string };
          return err(dsErr(body.error ?? `HTTP ${res.status}`));
        }
        return ok((await res.json()) as ArtifactView);
      } catch (e) {
        return err(dsErr(e instanceof Error ? e.message : "Network error"));
      }
    },

    async search(tenant, query) {
      return fetchJson<SearchResult[]>(
        `${baseUrl}/v1/${tenant}/search?q=${encodeURIComponent(query)}`,
      );
    },

    async traverse(tenant, from, relType, depth) {
      const params = new URLSearchParams();
      if (relType) params.set("type", relType);
      if (depth !== undefined) params.set("depth", String(depth));
      return fetchJson<ArtifactView[]>(`${baseUrl}/v1/${tenant}/traverse/${from}?${params}`);
    },

    async list(tenant, filter) {
      const params = new URLSearchParams();
      if (filter?.tags?.length) params.set("tags", filter.tags.join(","));
      const qs = params.toString();
      return fetchJson<ArtifactView[]>(`${baseUrl}/v1/${tenant}/list${qs ? `?${qs}` : ""}`);
    },

    async listTenants() {
      return fetchJson<TenantInfo[]>(`${baseUrl}/v1/tenants`);
    },

    async reload(tenant) {
      const result = await fetchJson<{ status: string; report: BuildReport }>(
        `${baseUrl}/v1/${tenant}/reload`,
        "POST",
      );
      if (!result.ok) return result;
      return ok(result.value.report);
    },

    async addKb(name, sourceDir) {
      const configResult = await configAddTenant(name, sourceDir);
      if (!configResult.ok) return err(dsErr(configResult.error.message));
      return ok(undefined);
    },

    async removeKb(name) {
      const configResult = await configRemoveTenant(name);
      if (!configResult.ok) return err(dsErr(configResult.error.message));
      return ok(undefined);
    },
  };
}

File: crates/pramana-tui/src/data_source.rs
use crate::error::TuiError;
use pramana_engine::{
    ArtifactView, BuildReport, ListFilter, SearchResult, TenantConfig, TenantInfo, TenantManager,
};

pub enum DataSource {
    Standalone(Box<TenantManager>),
    Daemon { port: u16 },
}

impl DataSource {
    pub fn mode_label(&self) -> &str {
        match self {
            DataSource::Standalone(_) => "standalone",
            DataSource::Daemon { .. } => "daemon",
        }
    }

    pub fn list_tenants(&self) -> Result<Vec<TenantInfo>, TuiError> {
        match self {
            DataSource::Standalone(tm) => Ok(tm.list_tenants()),
            DataSource::Daemon { port } => {
                let url = format!("http://localhost:{port}/v1/tenants");
                let body = daemon_get(&url)?;
                let tenants: Vec<TenantInfo> =
                    serde_json::from_str(&body).map_err(|e| TuiError::Http(e.to_string()))?;
                Ok(tenants)
            }
        }
    }

    pub fn list(
        &self,
        tenant: &str,
        filter: Option<&ListFilter>,
    ) -> Result<Vec<ArtifactView>, TuiError> {
        match self {
            DataSource::Standalone(tm) => {
                let reader = tm.reader(tenant)?;
                Ok(reader.list(filter)?)
            }
            DataSource::Daemon { port } => {
                let url = match filter.and_then(|f| f.tags.as_ref()) {
                    Some(tags) => format!(
                        "http://localhost:{port}/v1/{tenant}/list?tags={}",
                        tags.join(",")
                    ),
                    None => format!("http://localhost:{port}/v1/{tenant}/list"),
                };
                let body = daemon_get(&url)?;
                let artifacts: Vec<ArtifactView> =
                    serde_json::from_str(&body).map_err(|e| TuiError::Http(e.to_string()))?;
                Ok(artifacts)
            }
        }
    }

    pub fn get(&self, tenant: &str, slug: &str) -> Result<Option<ArtifactView>, TuiError> {
        match self {
            DataSource::Standalone(tm) => {
                let reader = tm.reader(tenant)?;
                Ok(reader.get(slug)?)
            }
            DataSource::Daemon { port } => {
                let url = format!("http://localhost:{port}/v1/{tenant}/get/{slug}");
                match daemon_get(&url) {
                    Ok(body) => {
                        let artifact: ArtifactView = serde_json::from_str(&body)
                            .map_err(|e| TuiError::Http(e.to_string()))?;
                        Ok(Some(artifact))
                    }
                    Err(_) => Ok(None),
                }
            }
        }
    }

    pub fn search(&self, tenant: &str, query: &str) -> Result<Vec<SearchResult>, TuiError> {
        match self {
            DataSource::Standalone(tm) => {
                let reader = tm.reader(tenant)?;
                Ok(reader.search(query, None)?)
            }
            DataSource::Daemon { port } => {
                let encoded = urlencoded(query);
                let url = format!("http://localhost:{port}/v1/{tenant}/search?q={encoded}");
                let body = daemon_get(&url)?;
                let results: Vec<SearchResult> =
                    serde_json::from_str(&body).map_err(|e| TuiError::Http(e.to_string()))?;
                Ok(results)
            }
        }
    }

    pub fn reload(&mut self, tenant: &str) -> Result<BuildReport, TuiError> {
        match self {
            DataSource::Standalone(tm) => Ok(tm.reload(tenant)?),
            DataSource::Daemon { port } => {
                let url = format!("http://localhost:{}/v1/{tenant}/reload", port);
                let body = daemon_post(&url)?;
                let report: BuildReport =
                    serde_json::from_str(&body).map_err(|e| TuiError::Http(e.to_string()))?;
                Ok(report)
            }
        }
    }

    pub fn add_kb(&mut self, name: &str, source_dir: &str) -> Result<BuildReport, TuiError> {
        match self {
            DataSource::Standalone(tm) => {
                let cfg = TenantConfig {
                    name: name.to_owned(),
                    source_dir: source_dir.to_owned(),
                };
                Ok(tm.mount(cfg)?)
            }
            DataSource::Daemon { .. } => Err(TuiError::General(
                "add-kb not supported in daemon mode (restart daemon to apply)".into(),
            )),
        }
    }

    pub fn remove_kb(&mut self, name: &str) -> Result<(), TuiError> {
        match self {
            DataSource::Standalone(tm) => Ok(tm.unmount(name)?),
            DataSource::Daemon { .. } => Err(TuiError::General(
                "remove-kb not supported in daemon mode (restart daemon to apply)".into(),
            )),
        }
    }

    pub fn check_daemon(port: u16) -> bool {
        ureq::get(&format!("http://localhost:{port}/v1/version"))
            .timeout(std::time::Duration::from_secs(1))
            .call()
            .is_ok()
    }
}

fn daemon_get(url: &str) -> Result<String, TuiError> {
    let resp = ureq::get(url)
        .timeout(std::time::Duration::from_secs(30))
        .call()
        .map_err(|e| TuiError::Http(e.to_string()))?;
    resp.into_string()
        .map_err(|e| TuiError::Http(e.to_string()))
}

fn daemon_post(url: &str) -> Result<String, TuiError> {
    let resp = ureq::post(url)
        .timeout(std::time::Duration::from_secs(30))
        .call()
        .map_err(|e| TuiError::Http(e.to_string()))?;
    resp.into_string()
        .map_err(|e| TuiError::Http(e.to_string()))
}

fn urlencoded(s: &str) -> String {
    let mut result = String::with_capacity(s.len() * 3);
    for b in s.bytes() {
        match b {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                result.push(b as char);
            }
            _ => {
                result.push('%');
                result.push(char::from(b"0123456789ABCDEF"[(b >> 4) as usize]));
                result.push(char::from(b"0123456789ABCDEF"[(b & 0xf) as usize]));
            }
        }
    }
    result
}
